File: providers/pricing.py
"""Shared LLM provider pricing table, used by the cost advisor and the
reconciliation job (drift detection needs to price gateway-observed
tokens the same way the gateway itself did).

This is the Python mirror of gateway/internal/provider/pricing.go --
same source, same verification date, same numbers. Keeping two files
instead of one shared-across-languages source is a deliberate,
documented trade (see docs/DECISIONS.md ADR-001 on the Go/Python split):
a single source of truth would need either a codegen step or a runtime
cross-language dependency, both heavier than re-stating one small table
in each language and reviewing them together on change.

Pricing verified 2026-09-10 against:
  - OpenAI: https://developers.openai.com/api/docs/pricing
  - Anthropic: https://platform.claude.com/docs/en/about-claude/pricing
    and https://platform.claude.com/docs/en/api/rate-limits (model roster
    cross-check)

Prices are USD per 1,000,000 tokens.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class PricePerMillion:
    input: float
    cached_input: float
    output: float


PRICING_TABLE: dict[str, PricePerMillion] = {
    # OpenAI
    "gpt-5.4": PricePerMillion(2.50, 0.25, 15.00),
    "gpt-5.4-mini": PricePerMillion(0.75, 0.075, 4.50),
    "gpt-5.4-nano": PricePerMillion(0.20, 0.02, 1.25),
    "gpt-5": PricePerMillion(1.25, 0.125, 10.00),
    "gpt-5-mini": PricePerMillion(0.25, 0.025, 2.00),
    "gpt-5-nano": PricePerMillion(0.05, 0.005, 0.40),
    "gpt-4.1": PricePerMillion(2.00, 0.50, 8.00),
    "gpt-4.1-mini": PricePerMillion(0.40, 0.10, 1.60),
    "gpt-4.1-nano": PricePerMillion(0.10, 0.025, 0.40),
    "gpt-4o": PricePerMillion(2.50, 1.25, 10.00),
    "gpt-4o-mini": PricePerMillion(0.15, 0.075, 0.60),
    "o1": PricePerMillion(15.00, 7.50, 60.00),
    "o3": PricePerMillion(2.00, 0.50, 8.00),
    "o3-mini": PricePerMillion(1.10, 0.55, 4.40),
    "o4-mini": PricePerMillion(1.10, 0.275, 4.40),
    # Anthropic
    "claude-fable-5-1": PricePerMillion(10.00, 0.25, 50.00),
    "claude-opus-5": PricePerMillion(5.00, 0.50, 25.00),
    "claude-sonnet-5": PricePerMillion(2.00, 0.20, 10.00),
    "claude-haiku-4-5": PricePerMillion(1.00, 0.10, 5.00),
    # Mock provider
    "mock-large": PricePerMillion(5.00, 0.50, 25.00),
    "mock-medium": PricePerMillion(1.00, 0.10, 5.00),
    "mock-small": PricePerMillion(0.15, 0.015, 0.60),
}


def normalize_model(model: str) -> str:
    if "/" in model:
        return model.split("/", 1)[1]
    return model


def price(model: str) -> PricePerMillion | None:
    return PRICING_TABLE.get(normalize_model(model))


def cost_usd(model: str, tokens_in: int, cached_input_tokens: int, tokens_out: int) -> float:
    p = price(model)
    if p is None:
        return 0.0
    uncached = max(tokens_in - cached_input_tokens, 0)
    return (
        uncached / 1_000_000 * p.input
        + cached_input_tokens / 1_000_000 * p.cached_input
        + tokens_out / 1_000_000 * p.output
    )


def model_family(model: str) -> str:
    m = normalize_model(model)
    if m.startswith("gpt-") or m.startswith("o1") or m.startswith("o3") or m.startswith("o4"):
        return "openai"
    if m.startswith("claude-"):
        return "anthropic"
    if m.startswith("mock-"):
        return "mock"
    return "unknown"


def cheaper_alternatives(model: str) -> list[str]:
    """Model IDs priced strictly below `model`'s output price, from the
    same provider family -- used by the advisor to only ever SUGGEST a
    switch that is actually cheaper (the gateway's policy engine
    independently re-verifies this before acting on it; see
    docs/DECISIONS.md ADR-003).
    """
    base = price(model)
    if base is None:
        return []
    family = model_family(model)
    return [
        name
        for name, p in PRICING_TABLE.items()
        if name != model and model_family(name) == family and p.output < base.output
    ]

File: gateway/internal/provider/pricing.go
package provider

import "strings"

// PricePerMillion holds per-1M-token USD pricing for one model. CachedInput
// is the price for cache-hit input tokens (0 if the provider doesn't offer
// prompt caching for that model).
type PricePerMillion struct {
	Input       float64
	CachedInput float64
	Output      float64
}

// pricingTable is embedded, current pricing verified 2026-09-10 against:
//   - OpenAI: https://developers.openai.com/api/docs/pricing
//   - Anthropic: https://platform.claude.com/docs/en/about-claude/pricing
//     and https://platform.claude.com/docs/en/api/rate-limits (model roster
//     cross-check)
//
// Prices are USD per 1,000,000 tokens. Keys are the bare model ID (no
// provider prefix) as sent in the request `model` field. See
// docs/DECISIONS.md for how staleness of this table is handled operationally
// (it is not; re-verifying it is a manual, documented process — this is a
// portfolio project, not a billing-critical production system, and a stale
// price here degrades cost estimates, it never bypasses budget enforcement,
// which is denominated in tokens first and dollars only as a convenience
// conversion).
var pricingTable = map[string]PricePerMillion{
	// OpenAI
	"gpt-5.4":      {Input: 2.50, CachedInput: 0.25, Output: 15.00},
	"gpt-5.4-mini": {Input: 0.75, CachedInput: 0.075, Output: 4.50},
	"gpt-5.4-nano": {Input: 0.20, CachedInput: 0.02, Output: 1.25},
	"gpt-5":        {Input: 1.25, CachedInput: 0.125, Output: 10.00},
	"gpt-5-mini":   {Input: 0.25, CachedInput: 0.025, Output: 2.00},
	"gpt-5-nano":   {Input: 0.05, CachedInput: 0.005, Output: 0.40},
	"gpt-4.1":      {Input: 2.00, CachedInput: 0.50, Output: 8.00},
	"gpt-4.1-mini": {Input: 0.40, CachedInput: 0.10, Output: 1.60},
	"gpt-4.1-nano": {Input: 0.10, CachedInput: 0.025, Output: 0.40},
	"gpt-4o":       {Input: 2.50, CachedInput: 1.25, Output: 10.00},
	"gpt-4o-mini":  {Input: 0.15, CachedInput: 0.075, Output: 0.60},
	"o1":           {Input: 15.00, CachedInput: 7.50, Output: 60.00},
	"o3":           {Input: 2.00, CachedInput: 0.50, Output: 8.00},
	"o3-mini":      {Input: 1.10, CachedInput: 0.55, Output: 4.40},
	"o4-mini":      {Input: 1.10, CachedInput: 0.275, Output: 4.40},

	// Anthropic
	"claude-fable-5-1": {Input: 10.00, CachedInput: 0.25, Output: 50.00},
	"claude-opus-5":    {Input: 5.00, CachedInput: 0.50, Output: 25.00},
	"claude-sonnet-5":  {Input: 2.00, CachedInput: 0.20, Output: 10.00},
	"claude-haiku-4-5": {Input: 1.00, CachedInput: 0.10, Output: 5.00},

	// Mock provider models, priced to mirror a cheap/mid/expensive spread
	// so the demo's "switch to a cheaper model" advisor suggestion has a
	// real cost delta to point at.
	"mock-large":  {Input: 5.00, CachedInput: 0.50, Output: 25.00},
	"mock-medium": {Input: 1.00, CachedInput: 0.10, Output: 5.00},
	"mock-small":  {Input: 0.15, CachedInput: 0.015, Output: 0.60},
}

// ErrUnknownModel-backed lookup: CostUSD returns 0 for unknown models
// rather than erroring, since pricing is advisory for budgeting purposes;
// the caller (budget package) treats an unpriced model as dollar-budget-
// exempt and falls back to pure token accounting, which still enforces a
// ceiling.
func Price(model string) (PricePerMillion, bool) {
	p, ok := pricingTable[NormalizeModel(model)]
	return p, ok
}

// CostUSD computes the dollar cost of a completed call. cachedInputTokens
// must be <= tokensIn.
func CostUSD(model string, tokensIn, cachedInputTokens, tokensOut int) float64 {
	p, ok := Price(model)
	if !ok {
		return 0
	}
	uncached := tokensIn - cachedInputTokens
	if uncached < 0 {
		uncached = 0
	}
	cost := float64(uncached)/1_000_000*p.Input +
		float64(cachedInputTokens)/1_000_000*p.CachedInput +
		float64(tokensOut)/1_000_000*p.Output
	return cost
}

// CheaperAlternatives returns model IDs priced strictly below `model`'s
// output price, from the same provider family (inferred by name prefix),
// cheapest-appropriate first. Used by the policy engine to evaluate a
// "switch_model" advisor suggestion against real pricing rather than
// trusting the LLM's claim that a model is cheaper.
func CheaperAlternatives(model string) []string {
	base, ok := Price(model)
	if !ok {
		return nil
	}
	family := modelFamily(model)
	var out []string
	for name, p := range pricingTable {
		if name == model {
			continue
		}
		if modelFamily(name) != family {
			continue
		}
		if p.Output < base.Output {
			out = append(out, name)
		}
	}
	return out
}

func modelFamily(model string) string {
	m := NormalizeModel(model)
	switch {
	case strings.HasPrefix(m, "gpt-") || strings.HasPrefix(m, "o1") || strings.HasPrefix(m, "o3") || strings.HasPrefix(m, "o4"):
		return "openai"
	case strings.HasPrefix(m, "claude-"):
		return "anthropic"
	case strings.HasPrefix(m, "mock-"):
		return "mock"
	default:
		return "unknown"
	}
}
